extensions/oidc/runtime/src/main/java/io/quarkus/oidc/Oidc.java

```diff
@@ -0,0 +1,50 @@
+package io.quarkus.oidc;
+
+import io.quarkus.oidc.common.runtime.config.OidcClientCommonConfig;
+import io.quarkus.oidc.runtime.OidcTenantConfig.ApplicationType;
+
+/**
+ * A CDI event that facilitates programmatic OIDC extension setup.
+ * Static OIDC tenants must be created in a synchronous manner from an observer method like in the example below:
+ *
+ * <pre>
+ * {@code
+ * public class OidcSetup {
+ *
+ *     void createStaticTenant(@Observes Oidc oidc) {
+ *         var defaultTenant = OidcTenantConfig.authServerUrl("https://oidc-provider-hostname/").build();
+ *         oidc.create(defaultTenant);
+ *     }
+ * }
+ * }
+ * </pre>
+ *
+ * The example above is equivalent to configuring {@code quarkus.oidc.auth-server-url=https://oidc-provider-hostname/}
+ * in the application.properties.
+ */
+public interface Oidc {
+
+    /**
+     * Creates static OIDC tenant. This method can also be used to create static tenants based on the tenants
+     * configured in the 'application.properties' file or any other configuration source.
+     *
+     * @param tenantConfig tenant config; must not be null
+     */
+    void create(OidcTenantConfig tenantConfig);
```

Comment:
Do we have any risk of blocked threads here since it is happening on the CDI event call ?

May be we can do something like this in the future: allow accepting async events, and `createAsync` methods will mark which tenant is being created right now, and then, when the request arrives and the tenant resolver determines that the current tenant is still being created then it will wait until it is initialized ?

Reply:
> Do we have any risk of blocked threads here since it is happening on the CDI event call ?

No, I don't think there is a risk. Remember that you also block threads when creating static tenant contexts, I don't think there is any difference here because it is almost same moment and definitely the same thread.

> May be we can do something like this in the future: allow accepting async events, and createAsync methods will mark which tenant is being created right now, and then, when the request arrives and the tenant resolver determines that the current tenant is still being created then it will wait until it is initialized ?

We definitely could do `createAsync`, then subscribe to all joined `Uni` and only create OIDC when is ready. There will need to be timeouts due, because while we can postpone creating OIDC ourselves, we cannot block others from creating it, because if there are incoming HTTP requests, it is going to be created anyway.

Async events alone won't do IMHO, because we can subscribe to the completion of the async events, but just because it is called "async", there isn't anything async about it. `@ObservesAsync` only says that other code after firing the event is not waiting till the event is processing. But it doesn't help you to for example doing async HTTP requests better than the normal event.

TL;DR; yes, I agree we can do better, but I'd probably wait till someone asks for do something about dynamic tenants. I am bit worried that if we pledge to postpone creating OIDC too much, will will start having little "bugs" or issues that this `oidc.create(...)` doesn't work for some combination with other extensions or whatever.

Reply:
btw I wrote in the Javadoc about "synchronous manner" but I only wanted users to create tenants immediately and not in the subscriber or on completion etc.